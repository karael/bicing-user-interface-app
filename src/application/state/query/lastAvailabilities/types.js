import { defineAction } from 'redux-define';

import { START, PENDING, CANCELLED, SUCCESS, FAILURE } from './../../stateConstants';

const domain = defineAction('lastAvailabilities');

export const FETCH_LIST = defineAction(
    'FETCH_LIST',
    [START, PENDING, CANCELLED, SUCCESS, FAILURE],
    domain
);
