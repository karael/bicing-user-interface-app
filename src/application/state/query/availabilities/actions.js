import { FETCH } from 'application/state/query/availabilities/types';

export const fetchStart = (
  itineraryStep,
  itineraryAt,
  periodStartAt,
  periodEndAt,
  interval,
  ...stationIds
) => ({
  error: false,
  meta: { isFetching: true, itineraryStep, itineraryAt },
  payload: {
    periodStartAt, periodEndAt, interval, stationIds,
  },
  type: FETCH.START,
});
export const fetchPending = itineraryStep => ({
  error: false,
  meta: { isFetching: true, itineraryStep },
  type: FETCH.PENDING,
});
export const fetchSuccess = (itineraryStep, data) => ({
  error: false,
  meta: { isFetching: false, itineraryStep },
  type: FETCH.SUCCESS,
  payload: data,
});
export const fetchFailure = (itineraryStep, error) => ({
  error: true,
  meta: { isFetching: false, itineraryStep },
  type: FETCH.FAILURE,
  payload: error,
});
