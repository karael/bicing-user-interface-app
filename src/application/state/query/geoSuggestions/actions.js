import { FETCH } from 'application/state/query/geoSuggestions/types';

export const fetchStart = () => ({
  error: false,
  meta: { isFetching: false },
  type: FETCH.START,
});
export const fetchPending = () => ({
  error: false,
  meta: { isFetching: true },
  type: FETCH.PENDING,
});
export const fetchSuccess = geoSuggestions => ({
  error: false,
  meta: { isFetching: false },
  type: FETCH.SUCCESS,
  payload: geoSuggestions,
});
export const fetchFailure = error => ({
  error: true,
  meta: { isFetching: false },
  type: FETCH.FAILURE,
  payload: error,
});
