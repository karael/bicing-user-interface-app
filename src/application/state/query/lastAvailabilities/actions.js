import * as Types from 'application/state/query/lastAvailabilities/types';

export const fetchListStart = () => ({ type: Types.FETCH_LIST.START, payload: {} });
export const fetchListPending = () => ({ type: Types.FETCH_LIST.PENDING, payload: {} });
export const fetchListCancelled = () => ({ type: Types.FETCH_LIST.CANCELLED, payload: {} });
export const fetchListSuccess = data => ({
    type: Types.FETCH_LIST.SUCCESS,
    payload: {
        data
    },
});
export const fetchListFailure = error => ({
    type: Types.FETCH_LIST.FAILURE,
    payload: {
        error,
    },
});
