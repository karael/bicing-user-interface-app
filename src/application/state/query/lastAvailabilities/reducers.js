import { createReducer, Types as ReduxSauceTypes } from 'reduxsauce'
import Immutable from 'seamless-immutable';

import * as Types from "./types";

export const INITIAL_STATE = Immutable({ error: false, isFetching: false, lastAvailabilities: [] });

export const fetchListStart = (state = INITIAL_STATE, action) => {
  const { isFetching } = action.payload;
  return { ...state, isFetching };
};

export const fetchListSuccess = (state = INITIAL_STATE, action) => {
  const { isFetching, data } = action.payload;
  return { ...state, isFetching, lastAvailabilities: data['hydra:member'] };
}

export const defaultHandler = (state, action) => {
  return { ...state };
}

export const HANDLERS = {
  [Types.FETCH_LIST.START]: fetchListStart,
  [Types.FETCH_LIST.SUCCESS]: fetchListSuccess,
  [ReduxSauceTypes.DEFAULT]: defaultHandler,
}

export default createReducer(INITIAL_STATE, HANDLERS);
