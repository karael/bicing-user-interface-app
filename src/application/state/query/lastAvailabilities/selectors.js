const selector = state => state.query.lastAvailabilities;

export const data = state => selector(state).data;
export const error = state => selector(state).error;
export const isFetching = state => selector(state).isFetching;

export const lastAvailabilityById = (state, id) => (data(state) === undefined
  ? undefined : data(state).find(model => model.id === id));

const property = (state, propertyKey, id) => (lastAvailabilityById(state, id) === undefined
  ? undefined : lastAvailabilityById(state, id)[propertyKey]);

export const statusByStationId = (state, stationId) => property(state, 'status', stationId);
export const availableBikeNumberByStationId = (state, stationId) => property(state, 'availableBikeNumber', stationId);
export const availableSlotNumberByStationId = (state, stationId) => property(state, 'availableSlotNumber', stationId);
