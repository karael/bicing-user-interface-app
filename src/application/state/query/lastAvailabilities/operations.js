import {
  fetchFailure,
  fetchPending,
  fetchSuccess,
} from 'application/state/query/lastAvailabilities/actions';
import { FETCH } from 'application/state/query/lastAvailabilities/types';
import HttpAvailabilityQuery from 'infrastructure/bicingApi/HttpAvailabilityQuery';
import Joi from 'joi';
import { call, put, takeLatest } from 'redux-saga/effects';
import { availabilitiesType } from 'domain/types/availabilityType';

export function* fetch() {
  yield put(fetchPending());

  try {
    const lastAvailabilities = yield call(HttpAvailabilityQuery.findAll);

    Joi.assert(lastAvailabilities, availabilitiesType);

    yield put(fetchSuccess(lastAvailabilities));
  } catch (exception) {
    yield put(fetchFailure(exception));
  }
}

export default function* operation() {
  yield takeLatest(FETCH.START, fetch);
}
