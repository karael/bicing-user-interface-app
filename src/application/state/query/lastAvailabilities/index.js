import reducer from 'application/state/query/lastAvailabilities/reducers';
import operation from 'application/state/query/lastAvailabilities/operations';
import * as actions from 'application/state/query/lastAvailabilities/actions';
import * as Types from 'application/state/query/lastAvailabilities/types';
import * as selectors from 'application/state/query/lastAvailabilities/selectors';

export {
    actions,
    Types,
    operation,
    selectors
};

export default reducer;
