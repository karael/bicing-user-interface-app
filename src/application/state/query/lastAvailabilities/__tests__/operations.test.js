import * as actions from 'application/state/query/lastAvailabilities/actions';
import operation, { fetch } from 'application/state/query/lastAvailabilities/operations';
import { FETCH } from 'application/state/query/lastAvailabilities/types';
import HttpAvailabilitiesQuery from 'infrastructure/bicingApi/HttpAvailabilitiesQuery';
import { expectSaga, testSaga } from 'redux-saga-test-plan';
import * as matchers from 'redux-saga-test-plan/matchers';
import { throwError } from 'redux-saga-test-plan/providers';

describe('application/state/query/lastAvailabilities/operations', () => {
  test('should wait for a fetch start event to fetch with operation()', () => {
    testSaga(operation)
      .next()
      .takeLatestEffect(FETCH.START, fetch);
  });

  test('should dispatch a fetchPending action with fetch() generator', () => {
    testSaga(fetch)
      .next()
      .put(actions.fetchPending());
  });

  test('should list expected lastAvailabilities with fetch() generator', () => {
    const fakeLastAvailabilities = [
      {
        '@id': '/api/last-availabilities-by-station/4028fe46-d10e-47df-9a66-f1960475af15',
        '@type': 'lastAvailabilitiesByStationView',
        id: '4028fe46-d10e-47df-9a66-f1960475af15',
        statedAt: '2018-09-19T12:50:03+02:00',
        availableBikeNumber: 1,
        availableSlotNumber: 25,
        status: 'OPENED',
      },
    ];

    return expectSaga(fetch)
      .provide([
        [matchers.call.fn(HttpAvailabilitiesQuery.find), fakeLastAvailabilities],
      ])
      .put(actions.fetchSuccess(fakeLastAvailabilities))
      .run();
  });

  test('should handle error when api call response does not contains expected schema type', () => {
    const fakeLastAvailabilitiesWithMissingRequiredProperties = [{ availableBikeNumber: 1 }];

    const action = {
      error: false,
      meta: { isFetching: true },
      payload: { latitude: 41.3244, longitude: 2.345, limit: 5000 },
      type: FETCH.START,
    };

    return expectSaga(fetch, action)
      .provide([
        [matchers.call.fn(
          HttpAvailabilitiesQuery.find,
        ),
        fakeLastAvailabilitiesWithMissingRequiredProperties,
        ],
      ])
      .run()
      .then((result) => {
        expect(result.toJSON()).toMatchSnapshot();
      });
  });

  test('should handle error when api call failed in fetch() generator', () => {
    const error = new Error('error_api_call');

    return expectSaga(fetch)
      .provide([
        [matchers.call.fn(HttpAvailabilitiesQuery.find), throwError(error)],
      ])
      .put(actions.fetchFailure(error))
      .run();
  });
});
