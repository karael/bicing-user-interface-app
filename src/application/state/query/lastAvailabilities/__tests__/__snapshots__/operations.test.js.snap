// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`application/state/query/lastAvailabilities/operations should handle error when api call response does not contains expected schema type 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/findAll",
      },
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": false,
          "meta": Object {
            "isFetching": true,
          },
          "type": "query/lastAvailabilities/FETCH_PENDING",
        },
        "channel": null,
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": true,
          "meta": Object {
            "isFetching": false,
          },
          "payload": Object {
            "_object": Array [
              Object {
                "availableBikeNumber": 1,
              },
            ],
            "annotate": "@@redux-saga-test-plan/json/function/<anonymous>",
            "details": Array [
              Object {
                "context": Object {
                  "key": "id",
                  "label": "id",
                },
                "message": "\\"id\\" is required",
                "path": Array [
                  0,
                  "id",
                ],
                "type": "any.required",
              },
            ],
            "isJoi": true,
            "name": "ValidationError",
          },
          "type": "query/lastAvailabilities/FETCH_FAILURE",
        },
        "channel": null,
      },
    },
  ],
}
`;
